### PR TITLE
[FSPE-6061] Created E2E test for Chi Popover component functionality

### PR DESCRIPTION
**Pull request by shintuviswambharan**
Create E2E test for Chi Popover component functionality

### CODE REVIEW

**Comment by allyourbotsarebelongtous** [comment]
Build was ABORTED. See http://jenkins.elasticbox.com/job/chi_pull_request/1331/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Your CHI Pull Request website is available for review at <a href="https://86584e9c4db74e5ef6467bc66a90ad8ac9b774b0-chi-pr.enterprise.ctl.io" target="_blank"> https://86584e9c4db74e5ef6467bc66a90ad8ac9b774b0-chi-pr.enterprise.ctl.io </a>

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/1332/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Your CHI Pull Request website is available for review at <a href="https://aa4c2bf32e842a531625686805afb34864f92e72-chi-pr.enterprise.ctl.io" target="_blank"> https://aa4c2bf32e842a531625686805afb34864f92e72-chi-pr.enterprise.ctl.io </a>

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/chi_pull_request/1335/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Your CHI Pull Request website is available for review at <a href="https://1dd52fd93f2d1fab1ad49789bed00341f6fe4ba0-chi-pr.enterprise.ctl.io" target="_blank"> https://1dd52fd93f2d1fab1ad49789bed00341f6fe4ba0-chi-pr.enterprise.ctl.io </a>

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/1336/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Your CHI Pull Request website is available for review at <a href="https://d14173ae4aee84caa934bbce41d3dbccc470c7c0-chi-pr.enterprise.ctl.io" target="_blank"> https://d14173ae4aee84caa934bbce41d3dbccc470c7c0-chi-pr.enterprise.ctl.io </a>

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/1338/ for more details.

**Review by janibolkvadze — CHANGES_REQUESTED** [review]
@shintuviswambharan Looks good! I think it covers well the main scenarios of Popover functionality. Some minor suggestions to improve performance and code linting. Thanks!

**Comment by allyourbotsarebelongtous** [comment]
Build was ABORTED. See http://jenkins.elasticbox.com/job/chi_pull_request/1344/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Build was ABORTED. See http://jenkins.elasticbox.com/job/chi_pull_request/1345/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Your CHI Pull Request website is available for review at <a href="https://8125a268f3477cf2daf4bc1e2babeea887a23bd3-chi-pr.enterprise.ctl.io" target="_blank"> https://8125a268f3477cf2daf4bc1e2babeea887a23bd3-chi-pr.enterprise.ctl.io </a>

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/1346/ for more details.

File: cypress/integration/chi-js/popover.spec.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        .first()
+        .click()
+        .find('+ .chi-popover')
+        .should('be.not.visible');
+    });
+  });
+
+  describe('Popover Positioning and arrow should be matched ', () => {
+    it('Popover Positioning and arrow should work in accordance', () => {
+      this.chidata.popperPositions.forEach(position => {
```
[the remark this answers is not in the document]
Comment (shintuviswambharan) [reply cypress/integration/chi-js/popover.spec.js:26]: Thank you

File: cypress/integration/chi-js/popover.spec.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    cy.fixture('chidata.json').then(data => {
+      this.chidata = data;
+    });
+  });
+  describe('Popover should be shown  when  button  clicked ', function() {
+    it('Popover should shown on button clicked', () => {
+      cy.get('button.chi-button')
+        .first()
+        .click()
+        .find('+ .chi-popover')
```
[the remark this answers is not in the document]
Comment (shintuviswambharan) [reply cypress/integration/chi-js/popover.spec.js:12]: Updated in latest commit

File: cypress/integration/chi-js/popover.spec.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        .should('be.visible')
+        .should('have.class', '-active');
+    });
+    it('Popover should hide  on button clicked twice', () => {
+      cy.get('button.chi-button')
```
[the remark this answers is not in the document]
Comment (shintuviswambharan) [reply cypress/integration/chi-js/popover.spec.js]: Updated in latest commit

File: cypress/integration/chi-js/popover.spec.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
+describe('Popover', function() {
+  before(() => {
+    cy.visit('http://localhost:8000/tests/js/popover.html');
```
[the remark this answers is not in the document]
Comment (shintuviswambharan) [reply cypress/integration/chi-js/popover.spec.js]: Updated in latest commit

File: test/chi/js/popover.pug [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  button#show-popover-3.chi-button Show
+  button#hide-popover-3.chi-button Hide
+  button#toggle-popover-3.chi-button Toggle
+  button#popover-3.chi-button(data-popover-content='<div class="chi-popover__content">Foo</div>') Popover
+  script.
+    var popover = chi.popover(
+    document.getElementById('popover-3'),
+    {
+    trigger: 'manual',
+    preventAutoHide: true
```
[the remark this answers is not in the document]
Comment (shintuviswambharan) [reply test/chi/js/popover.pug]: Updated in latest commit

File: cypress/integration/chi-js/popover.spec.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    });
+  });
+
+  describe('Popover Positioning and arrow should be matched ', () => {
+    it('Popover Positioning and arrow should work in accordance', () => {
+      this.chidata.popperPositions.forEach(position => {
+        const getValue = `.test-more-${position}`;
+        cy.get(getValue)
+          .find('button.chi-button')
+          .scrollIntoView()
```
[the remark this answers is not in the document]
Comment (shintuviswambharan) [reply cypress/integration/chi-js/popover.spec.js]: Updated in latest commit

File: test/chi/js/popover.pug [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    }
+    );
+    document.getElementById('show-popover-3').addEventListener('click', function(e) {
+    popover.show()
+    });
+    document.getElementById('hide-popover-3').addEventListener('click', function(e) {
+    popover.hide()
+    });
+    document.getElementById('toggle-popover-3').addEventListener('click', function(e) {
+    popover.toggle()
```
[the remark this answers is not in the document]
Comment (shintuviswambharan) [reply test/chi/js/popover.pug]: Updated in latest commit

File: test/chi/js/popover.pug [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    {
+    trigger: 'manual',
+    preventAutoHide: true
+    }
+    );
+    document.getElementById('show-popover-3').addEventListener('click', function(e) {
+    popover.show()
+    });
+    document.getElementById('hide-popover-3').addEventListener('click', function(e) {
+    popover.hide()
```
[the remark this answers is not in the document]
Comment (shintuviswambharan) [reply test/chi/js/popover.pug]: Updated in latest commit

**Comment by allyourbotsarebelongtous** [comment]
Your CHI Pull Request website is available for review at <a href="https://15edb7112f0f2b20ccc36b68ee792d8b824db61a-chi-pr.enterprise.ctl.io" target="_blank"> https://15edb7112f0f2b20ccc36b68ee792d8b824db61a-chi-pr.enterprise.ctl.io </a>

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/1348/ for more details.

**Comment by shintuviswambharan** [comment]
@janibolkvadze , I have updated test cases based on the review comments . Could you please review the same

File: cypress/integration/chi-js/popover.spec.js [hunk 9]
```diff
@@ -0,0 +1,107 @@
+describe('Popover', function() {
+  before(() => {
+    cy.visit('/tests/js/popover.html');
+    cy.fixture('chidata.json').then(data => {
+      this.chidata = data;
+    });
+  });
```
Comment (janibolkvadze) [inline cypress/integration/chi-js/popover.spec.js:7]:
```suggestion
  });

```

File: cypress/integration/chi-js/popover.spec.js [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+  describe('Popover should be shown  when  button  clicked ', function() {
+    it('Popover should shown on button clicked', () => {
+      cy.get('[data-cy="show-popover"]')
+        .click()
+        .find('+ .chi-popover')
+        .should('be.visible')
+        .should('have.class', '-active');
+    });
```
Comment (janibolkvadze) [inline cypress/integration/chi-js/popover.spec.js:15]:
```suggestion
    });

```

File: cypress/integration/chi-js/popover.spec.js [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+  describe('Autohide and NoAutohide should work', function() {
+    it('popover should hide when click out side when preventAutoHide set to false', () => {
+      cy.get('[data-cy="auto-hide"]')
+        .find('[data-cy="popover-6"]')
+        .click()
+        .parent()
+        .find('.chi-popover#chi-popover-13')
+         .should('have.class', '-active')
+         .click(500, 0, {force: true})
+         .find('.chi-popover#chi-popover-13')
```
Comment (janibolkvadze) [inline cypress/integration/chi-js/popover.spec.js:46]:
```suggestion
       .should('have.class', '-active')
       .click(500, 0, {force: true})
       .find('.chi-popover#chi-popover-13')
```

File: cypress/integration/chi-js/popover.spec.js [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+        .get('.parent')
+        .find('.chi-popover#chi-popover-14')
+        .should('have.class', '-active')
+        .and('match', `[aria-hidden="false"]`);
+    });
+  });
+
+  describe('Hide, Toggle, Show events should work ', function() {
+    it('Hide, show, toggle should work', () => {
+      //trigger show
```
Comment (janibolkvadze) [inline cypress/integration/chi-js/popover.spec.js:67]:
```suggestion
```

File: test/chi/js/popover.pug [hunk 13]
```diff
@@ -0,0 +1,67 @@
+---
+title: Popover
+---
+script(src='/js/chi.js')
+
+- var moreSides = ["top", "top-start", "top-end", "right", "right-start", "right-end", "bottom", "bottom-start", "bottom-end", "left", "left-start", "left-end"];
+
+// Method 1 - Data Target
```
Comment (janibolkvadze) [inline test/chi/js/popover.pug:8]: Let's remove all the comments, please.

File: test/chi/js/popover.pug [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+button#popover-1b.chi-button(data-target='#popover-in-html',data-cy='show-popover') Popover
+// Put the following code anywere in the DOM, e.g. at the end of the DOM
+section#popover-in-html.chi-popover(aria-modal='true', role='dialog', aria-label='Popover title')
+  header.chi-popover__header
+    h2.chi-popover__title Popover title
+  .chi-popover__content
+    p.chi-popover__text Popover content
+script.
```
Comment (janibolkvadze) [inline test/chi/js/popover.pug:16]: Let's consolidate `chi.popover()` inside one `script.` on the button of the pug file

File: cypress/integration/chi-js/popover.spec.js [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+        .should('have.class', '-active')
+        .click(500, 0, {force: true})
+        .get('.parent')
+        .find('.chi-popover#chi-popover-14')
+        .should('have.class', '-active')
+        .and('match', `[aria-hidden="false"]`);
+    });
+  });
+
+  describe('Hide, Toggle, Show events should work ', function() {
```
Comment (janibolkvadze) [inline cypress/integration/chi-js/popover.spec.js:65]:
```suggestion
  describe('Hide, Toggle, Show methods should work ', function() {
```

Comment (janibolkvadze) [reply cypress/integration/chi-js/popover.spec.js:65]: The click events call the .show(), .hide(), .toggle() methods.
Regarding events: Can we add test cases to check event emission of:
**chi.popover.show
chi.popover.hide** ?
Popover is supposed to emit these events when show and hide are triggered. Would be nice to test that as well.

**Review by janibolkvadze — CHANGES_REQUESTED** [review]

**Comment by allyourbotsarebelongtous** [comment]
Your CHI Pull Request website is available for review at <a href="https://0a7473cc6b4186bd7d6b49317155ae2b41a5d073-chi-pr.enterprise.ctl.io" target="_blank"> https://0a7473cc6b4186bd7d6b49317155ae2b41a5d073-chi-pr.enterprise.ctl.io </a>

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/1358/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Your CHI Pull Request website is available for review at <a href="https://ff7430a4789a8140d4c34183fe78e11a22a62720-chi-pr.enterprise.ctl.io" target="_blank"> https://ff7430a4789a8140d4c34183fe78e11a22a62720-chi-pr.enterprise.ctl.io </a>

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/1359/ for more details.

File: test/chi/js/popover.pug (hunk 13 above)
Comment (shintuviswambharan) [reply test/chi/js/popover.pug]: updated in latest commit

**Comment by allyourbotsarebelongtous** [comment]
Your CHI Pull Request website is available for review at <a href="https://3172c60d3bef9bbacb8a2acfc80cfd582ee7852b-chi-pr.enterprise.ctl.io" target="_blank"> https://3172c60d3bef9bbacb8a2acfc80cfd582ee7852b-chi-pr.enterprise.ctl.io </a>

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/1387/ for more details.

File: test/chi/js/popover.pug [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+section#popover-in-html.chi-popover(aria-modal='true', role='dialog', aria-label='Popover title')
+  header.chi-popover__header
+    h2.chi-popover__title Popover title
+  .chi-popover__content
+    p.chi-popover__text Popover content
+
+each side in moreSides
+    div(class=`test-more-${side} -position--relative`,data-cy=`test-more-${side}`)
+        button.chi-button.popover-btn(data-position=`${side}`, data-popover-content='<header class="chi-popover__header"><h2 class="chi-popover__title">Popover positioning</h2></header><div class="chi-popover__content"><p class="chi-popover__text">Popover Content.</p></div>')=`${side} popover`
```
Comment (janibolkvadze) [inline test/chi/js/popover.pug:18]:
```suggestion
  div(class=`test-more-${side} -position--relative`,data-cy=`test-more-${side}`)
    button.chi-button.popover-btn(data-position=`${side}`, data-popover-content='<header class="chi-popover__header"><h2 class="chi-popover__title">Popover positioning</h2></header><div class="chi-popover__content"><p class="chi-popover__text">Popover Content.</p></div>')=`${side} popover`
```

File: test/chi/js/popover.pug [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+.parent(data-cy='auto-hide')
+
```
Comment (janibolkvadze) [inline test/chi/js/popover.pug:21]:
```suggestion
```

File: test/chi/js/popover.pug [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+  button#popover-6.chi-button(data-cy='popover-6',data-popover-content='<header class="chi-popover__header"><h2 class="chi-popover__title">Popover title</h2></header><div class="chi-popover__content"><p class="chi-popover__text">Click outside. I will disappear!</p></div>') Auto hide
+  script.
```
Comment (janibolkvadze) [inline test/chi/js/popover.pug:23]: Can we move all the JS to the bottom of the file?
Let's consolidate all the scripts together, please.

File: cypress/integration/chi-js/popover.spec.js [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+        .should('be.not.visible');
+    });
+
+  });
+
+  describe('Popover Positioning and arrow should be matched ', () => {
+    it('Popover Positioning and arrow should work in accordance', () => {
+      this.chidata.popperPositions.forEach(position => {
+        const getValue = `[data-cy="test-more-${position}"]`;
+        cy.get(getValue)
```
Comment (janibolkvadze) [inline cypress/integration/chi-js/popover.spec.js:31]:
```suggestion

        cy.get(getValue)
```

File: cypress/integration/chi-js/popover.spec.js [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+        .find('#chi-popover-2.chi-popover')
+        .should('have.class', '-active')
+        .should('match', `[aria-hidden="false"]`)
+        .parent()
+        .click()
+        .find('#chi-popover-2.chi-popover')
+        .should('not.have.class', '-active')
+        .should('match', `[aria-hidden="true"]`);
+    });
+
```
Comment (janibolkvadze) [inline cypress/integration/chi-js/popover.spec.js:97]:
```suggestion
```

File: cypress/integration/chi-js/popover.spec.js [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+      cy.get('[data-cy="dynamic-content"]')
+        .click()
+        .parent()
+        .find('#chi-popover-16')
+        .should('have.class', '-active')
+        .get('h4')
+        .contains('This is data-popover-content content')
+        .should('exist');
+    });
+
```
Comment (janibolkvadze) [inline cypress/integration/chi-js/popover.spec.js:111]:
```suggestion
```

**Review by janibolkvadze — COMMENTED** [review]
Just a couple of minor tweakings and it will be perfect. Thank you!

**Comment by allyourbotsarebelongtous** [comment]
Build was ABORTED. See http://jenkins.elasticbox.com/job/chi_pull_request/1393/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Your CHI Pull Request website is available for review at <a href="https://aedb1fbb7203634ec3028d09e2ac69ae7e2c7359-chi-pr.enterprise.ctl.io" target="_blank"> https://aedb1fbb7203634ec3028d09e2ac69ae7e2c7359-chi-pr.enterprise.ctl.io </a>

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/1398/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/chi_pull_request/1403/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/chi_pull_request/1416/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/chi_pull_request/1417/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/chi_pull_request/1418/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Your CHI Pull Request website is available for review at <a href="https://f3190f0e4495737ead7864506fa2d22a7ddd5b85-chi-pr.enterprise.ctl.io" target="_blank"> https://f3190f0e4495737ead7864506fa2d22a7ddd5b85-chi-pr.enterprise.ctl.io </a>

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/1419/ for more details.

**Comment by shintuviswambharan** [comment]
@janibolkvadze , Made the changes based on the review comments. Thank you

**Review by janibolkvadze — APPROVED** [review]
Awesome, thank you!

**Comment by allyourbotsarebelongtous** [comment]
The following instances are being terminated: https://ops.elasticbox.com/services/instances/i-akwsq6, https://ops.elasticbox.com/services/instances/i-3lqthe, https://ops.elasticbox.com/services/instances/i-510zda, https://ops.elasticbox.com/services/instances/i-wehf79, https://ops.elasticbox.com/services/instances/i-kolf5c, https://ops.elasticbox.com/services/instances/i-vqsu5t, https://ops.elasticbox.com/services/instances/i-jwg94i, https://ops.elasticbox.com/services/instances/i-qwxa1r, https://ops.elasticbox.com/services/instances/i-fz9ec9, https://ops.elasticbox.com/services/instances/i-txh3bs, https://ops.elasticbox.com/services/instances/i-7ie86i

**Pull request merged**